File: Leetcode/525. Contiguous Array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int,int> m;
        int sum = 0, mx = 0;
        for(int i=0; i<nums.size(); i++) {
            if(nums[i]) sum++;
            else sum--;
            
            if(sum == 0) mx = max(mx, i+1);
            else if(m.find(sum) != m.end()) {
                mx = max(mx, i - m[sum]);
            } else m[sum] = i;
        }
        
        return mx;
    }
};

File: Leetcode/Rust/525. Contiguous Array.rs
use std::collections::HashMap;
use std::cmp;

impl Solution {
    pub fn find_max_length(nums: Vec<i32>) -> i32 {
        let mut m = HashMap::new();
        let (mut sum, mut mx) = (0, 0);
        
        for i in 0..nums.len() {
            sum += if nums[i] == 1 { 1 } else { -1 };
            
            if sum == 0 {
                mx = cmp::max(mx, i+1);
            } else if m.contains_key(& sum) {
                mx = cmp::max(mx, i - m.get(& sum).unwrap());
            } else {
                m.insert(sum, i);
            }
        }
        
        mx as i32
    }
}
